binding/dotnet/PvSpeaker/PvSpeaker.cs

```diff
@@ -0,0 +1,441 @@
+            {
+                throw new PvSpeakerInvalidArgumentException($"Sample rate of {sampleRate} is invalid - must be greater than 0.");
+            }
+
+            if (bitsPerSample != 8 && bitsPerSample != 16 && bitsPerSample != 24 && bitsPerSample != 32)
+            {
+                throw new PvSpeakerInvalidArgumentException($"Bits per sample of {bitsPerSample} is invalid - must be 8, 16, 24, or 32.");
+            }
+
+            if (bufferSizeSecs <= 0)
+            {
+                throw new PvSpeakerInvalidArgumentException($"Buffer size (in seconds) of {bufferSizeSecs} is invalid - must be greater than 0.");
+            }
+
+            if (deviceIndex < -1)
+            {
+                throw new PvSpeakerInvalidArgumentException($"Device index of {deviceIndex} is invalid - must be greater than -1.");
+            }
+
+            PvSpeakerStatus status = pv_speaker_init(sampleRate, bitsPerSample, bufferSizeSecs, deviceIndex, out _libraryPointer);
+            if (status != PvSpeakerStatus.SUCCESS)
+            {
+                throw PvSpeakerStatusToException(status);
+            }
+
+            SampleRate = sampleRate;
+            BitsPerSample = bitsPerSample;
+            BufferSizeSecs = bufferSizeSecs;
+            SelectedDevice = Marshal.PtrToStringAnsi(pv_speaker_get_selected_device(_libraryPointer));
+            Version = Marshal.PtrToStringAnsi(pv_speaker_version());
+        }
+
+        /// <summary>
+        /// Starts the audio output device. Should be called before making any calls to `Write()`, `Flush()`, or `Stop()`.
+        /// </summary>
+        public void Start()
+        {
+            PvSpeakerStatus status = pv_speaker_start(_libraryPointer);
+            if (status != PvSpeakerStatus.SUCCESS)
+            {
+                throw PvSpeakerStatusToException(status);
+            }
+        }
+
+        /// <summary>
+        /// Stops the audio output device. Should only be called after a successful call to `Start()`.
+        /// </summary>
+        public void Stop()
+        {
+            PvSpeakerStatus status = pv_speaker_stop(_libraryPointer);
+            if (status != PvSpeakerStatus.SUCCESS)
+            {
+                throw PvSpeakerStatusToException(status);
+            }
+        }
+
+        /// <summary>
+        /// Synchronous call to write PCM data to the internal circular buffer for audio playback.
+        /// Only writes as much PCM data as the internal circular buffer can currently fit, and returns
+        /// the number of samples that were successfully written. Call between `Start()` and `Stop()`.
+        /// </summary>
+        /// <param name="pcm">PCM data to be played.</param>
+        /// <returns>Number of samples that were successfully written.</returns>
+        public int Write(byte[] pcm)
+        {
+            GCHandle pinnedArray = GCHandle.Alloc(pcm, GCHandleType.Pinned);
+            IntPtr pcmPtr = pinnedArray.AddrOfPinnedObject();
```

Comment:
why do we do this? may want to add a comment that let's the reader know why we do this

Reply:
pvspeaker `c` accepts an `int8_t` pointer to the PCM and handles it manually based on the bits per sample so it can handle 8, 16, 24, 32 bit data

did you mean to explain the above to the user, or just that those lines are to get a pointer from the `byte[]`?

Reply:
I was commenting on the two highlighted lines, where the pcm is converted to an IntPtr. I just didn't understand the pinning and thought others might not either. After looking it up though, it's not too weird so I guess it can go without a comment. If you do leave a comment it could be as simple as `Convert managed pcm buffer to unmanaged memory` or something